internal/providers/pluginfw/resources/catalog/resource_function.go

```diff
@@ -0,0 +1,226 @@
+package catalog
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/retries"
+	"github.com/databricks/databricks-sdk-go/service/catalog"
+	"github.com/databricks/terraform-provider-databricks/common"
+	pluginfwcommon "github.com/databricks/terraform-provider-databricks/internal/providers/pluginfw/common"
+	pluginfwcontext "github.com/databricks/terraform-provider-databricks/internal/providers/pluginfw/context"
+	"github.com/databricks/terraform-provider-databricks/internal/providers/pluginfw/converters"
+	"github.com/databricks/terraform-provider-databricks/internal/providers/pluginfw/tfschema"
+	"github.com/databricks/terraform-provider-databricks/internal/service/catalog_tf"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+)
+
+const resourceName = "function"
+
+var _ resource.ResourceWithConfigure = &FunctionResource{}
+
+func ResourceFunction() resource.Resource {
+	return &FunctionResource{}
+}
+
+func waitForFunction(ctx context.Context, w *databricks.WorkspaceClient, funcInfo *catalog.FunctionInfo) diag.Diagnostics {
+	const timeout = 5 * time.Minute
+
+	result, err := retries.Poll[catalog.FunctionInfo](ctx, timeout, func() (*catalog.FunctionInfo, *retries.Err) {
+		attempt, err := w.Functions.GetByName(ctx, funcInfo.FullName)
+		if err != nil {
+			if apierr.IsMissing(err) {
+				return nil, retries.Continue(fmt.Errorf("function %s is not yet available", funcInfo.FullName))
+			}
+			return nil, retries.Halt(fmt.Errorf("failed to get function: %s", err))
+		}
+		return attempt, nil
+	})
+
+	if err != nil {
+		return diag.Diagnostics{diag.NewErrorDiagnostic("failed to create function", err.Error())}
+	}
+
+	*funcInfo = *result
+	return nil
+}
+
+type FunctionResource struct {
+	Client *common.DatabricksClient
+}
+
+func (r *FunctionResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = pluginfwcommon.GetDatabricksProductionName(resourceName)
+}
+
+func (r *FunctionResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	attrs, blocks := tfschema.ResourceStructToSchemaMap(catalog_tf.FunctionInfo{}, func(c tfschema.CustomizableSchema) tfschema.CustomizableSchema {
+		c.SetRequired("name")
+		c.SetRequired("catalog_name")
+		c.SetRequired("schema_name")
+		c.SetRequired("input_params")
+		c.SetRequired("data_type")
+		c.SetRequired("routine_body")
+		c.SetRequired("routine_defintion")
+		c.SetRequired("language")
```

Comment:
`full_data_type` marked as `required` in API spec: https://docs.databricks.com/api/workspace/functions/create but I'm not sure if it's really required. We need to check with UC team

Reply:
similarly for [is_null_call](https://docs.databricks.com/api/workspace/functions/create#function_info-is_null_call) and [is_deterministic](https://docs.databricks.com/api/workspace/functions/create#function_info-is_deterministic)